tutorial/testing.py

```diff
@@ -0,0 +1,297 @@
+    All the rules apply as in your app file.
+
+    * #4 We normally start the test by calling the `start_server` API
+    from dash_duo. Several actions implicitly happen under the hood:
+
+        1. The defined app is hosted inside a light python `threading.Thread`.
+        2. A selenium webdriver is initialized and navigates to the
+        local server URL using `server_url`.
+        3. We first wait until the flask server is responsive to an HTTP
+        request, and then make sure the Dash app is full rendered inside
+        the browser.
+
+    * #5 A test case is composed of preparation, actions, and checkpoints.
+    Both #5 and #6 are doing the same check in this example; we are expecting
+    that the defined `Div` component's text is identical to `children`. #5 will
+    wait for the expected state to be reached within a 4 seconds timeout. It's
+    a safer way to write the action steps when you are doing an element check
+    related to callbacks, as it normally happens under Dash context:
+    the element is already present in the DOM, but not necessarily the props.
+
+    * #6 The `find_element` API call has an implicitly global timeout of two
+    seconds set at the driver level, i.e. the driver waits at most two seconds
+    to find the element by the locator, **HOWEVER** it will compare the text
+    as soon as the driver returns an element. Also note that the API
+    `find_element('#nully-wrapper')` is just a shortcut to a more tedious
+    version `driver.find_element_by_css_selector('#nully-wrapper')`.
+
+    * #7 Unlike `unittest`, `pytest` uses the native python
+    [`assert`](https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement)
+    statement to do assertions. It's good practice to expose your acceptance
+    criteria directly in the test case rather than wrapping the assert inside
+    another helper API. By looking at the test case, reviewers and maintainers
+    should be able to easily figure out the purpose of the test by
+    the test title, the app definition, the actions, and the checkpoints.
+
+    * #8 We use [Percy](https://percy.io/) as our *Visual Regression Testing*
+    tool. It's a good alternative to assertions when your checkpoint is
+    about the graphical aspects of a Dash app, such as the whole layout or a
+    `dcc.Graph` component. We integrate the Percy service
+    with a `PERCY_TOKEN` variable,  so the regression result is only
+    available in Plotly CircleCI setup.
+
+    ## Fixtures
+
+    To avoid accidental name collision with other pytest plugins,
+    all Dash test fixtures start with the prefix `dash_`.
+
+    - dash_duo
+
+    The default fixture for Dash integration tests, it contains a
+    `thread_server` and a webdriver wrapped with high-level Dash testing APIs.
+
+    - dash_br
+
+    A standalone webdriver wrapped with high-level Dash testing APIs. This is
+    suitable for testing a Dash app in a deployed environment, i.e. one in
```

Comment:
```suggestion
    suitable for testing a Dash app in a deployed environment, i.e., when
```
removing "which" afterwards